google/cloud/pubsublite/internal/partition_publisher.h

```diff
@@ -0,0 +1,322 @@
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace pubsublite_internal {
+
+using google::cloud::pubsublite::v1::Cursor;
+using google::cloud::pubsublite::v1::MessagePublishRequest;
+using google::cloud::pubsublite::v1::MessagePublishResponse;
+using google::cloud::pubsublite::v1::PubSubMessage;
+
+using std::deque;
+
+class PartitionPublisherImpl : public Publisher<Cursor> {
+ public:
+  explicit PartitionPublisherImpl(
+      absl::FunctionRef<std::unique_ptr<ResumableAsyncStreamingReadWriteRpc<
+          MessagePublishRequest, MessagePublishResponse>>(
+          StreamInitializer<MessagePublishRequest, MessagePublishResponse>)>
+          resumable_stream_factory,
+      BatchingSettings const batching_settings,
+      std::unique_ptr<AlarmRegistryInterface> alarm, size_t alarm_period_ms)
+      : batching_settings_{batching_settings},
+        resumable_stream_{resumable_stream_factory(initializer_)},
+        cancel_token_{alarm->RegisterAlarm(absl::Milliseconds(alarm_period_ms),
+                                           alarm_)} {}
+
+  future<Status> Start() override {
+    promise<void> root;
+    future<void> root_future = root.get_future();
+    future<Status> return_future;
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      return_future = root_future.then(ChainFuture(resumable_stream_->Start()))
+                          .then([this](future<Status> start_status) {
+                            {
+                              std::lock_guard<std::mutex> g{mu_};
+                              if (shutdown_) return start_status.get();
+                              shutdown_ = true;
+                              DestroyCancelToken();
+                            }
+                            Status status = start_status.get();
+                            SatisfyOutstandingMessages(status);
+                            return status;
+                          });
+    }
+    root.set_value();
+    Read();
+    return return_future;
+  }
+
+  future<StatusOr<Cursor>> Publish(PubSubMessage m) override {
+    future<StatusOr<Cursor>> message_future;
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      if (shutdown_) return make_ready_future(StatusOr<Cursor>(Status()));
+      promise<StatusOr<Cursor>> message_promise;
+      message_future = message_promise.get_future();
+      unbatched_messages_with_futures_.emplace_back(
+          MessageWithFuture{std::move(m), std::move(message_promise)});
+    }
+    return message_future;
+  }
+
+  void Flush() override {
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      if (writing_ || shutdown_) return;
+      writing_ = true;
+    }
+    Rebatch(false);
+    WriteBatches();
+  }
+
+  future<void> Shutdown() override {
+    promise<void> root;
+    future<void> root_future = root.get_future();
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      if (shutdown_) return make_ready_future();
+      shutdown_ = true;
+      DestroyCancelToken();
```

Comment:
You can't do this under lock: this will block until the alarm downcall is done running, instead, just do this outside of the lock before root.set_value and unmark this as guarded_by (it's not racy) and replace it with `cancel_token_ = nullptr;`